### DISCUSSION TITLE
Parser not recovering correctly

### DISCUSSION THREAD

**Discussion by auctumnus** [zesterer/chumsky#395] [Q&A] [answered]
Hello! I'm having a somewhat confusing result from a parser I've made. A minimal repro would be somewhat confusing, so here's a [repo with the affected code](https://github.com/auctumnus/chumsky-issue).

At the moment, I would expect that from `ast_element`, if `rule` fails to parse, then `cat_edit` would be ran to parse the line. However, the parsing seems to just fail if the first of the two fails - you can see this by switching around the order of the parsers in the `choice` statement.

My only other thought would be that it fails to rewind correctly because both can start with similar sequences, but I'm not sure how otherwise to parse this - with the 0.x versions, maybe `then_with` would work, but `then_with_ctx` looks entirely different.

**Comment by zesterer** [comment]
The example you gave is quite large and explicit. Do you think you could comment out the bits that don't relate to the error and come up with a minimal failing example? That's usually the best way to go about solving issues like this.

**Reply by auctumnus** [reply]
Apologies, I've now reduced the test case and found something which fixes it, but it seems like a really confusing case that I can't explain. [At this line in `main.rs`](https://github.com/auctumnus/chumsky-issue/blob/main/src/main.rs#L38), adding `at_least(1)` fixes the issue. Is this just a pathological case for the error message?

**Reply by zesterer** [reply, marked as answer]
`separated_by`, like `repeated`, can accept no inputs at all and hence will always succeed.

**Reply by auctumnus** [reply]
I still don't really get why the error message was the particular thing it was, but I figure I'm just missing something and I'll realize it afterwards.
